packages/odf/components/s3-browser/upload-objects/uploads.ts

```diff
@@ -0,0 +1,82 @@
+import * as React from 'react';
+import { Upload } from '@aws-sdk/lib-storage';
+import { S3Commands } from '@odf/shared/s3';
+import * as _ from 'lodash-es';
+import { UploadProgress, UploadStatus } from './types';
+
+export function* uploadFile(
+  files: File[],
+  client: S3Commands,
+  bucketName: string,
+  setUploadProgress: React.Dispatch<React.SetStateAction<UploadProgress>>
+): Generator<unknown, unknown, ReturnType<typeof Upload.prototype.done>[]> {
+  const performUploadPromise = (
+    file: File
+  ): ReturnType<typeof Upload.prototype.done> => {
+    const uploader = new Upload({
+      client: client,
+      params: {
+        Bucket: bucketName,
+        Key: file.name,
+        Body: file,
+      },
+      queueSize: 4,
+    });
+    uploader.on('httpUploadProgress', (progress) => {
+      const isComplete = progress.total === progress.loaded;
+      setUploadProgress((prev) => ({
+        ...prev,
+        [progress.Key]: {
+          startTime: prev[progress.Key].startTime ?? Date.now(),
+          name: progress.Key,
+          total: progress.total,
+          loaded: progress.loaded,
+          abort: () => {
+            uploader.abort();
+            setUploadProgress((previous) => {
+              const clonedPrev = _.cloneDeep(previous);
+              clonedPrev[progress.Key].uploadState = UploadStatus.UPLOAD_FAILED;
+              return clonedPrev;
+            });
+          },
+          uploadState: isComplete
+            ? UploadStatus.UPLOAD_COMPLETE
+            : UploadStatus.UPLOAD_START,
+        },
+      }));
+    });
+    return uploader.done();
+  };
+  try {
+    const allUploadPromise = files.map((file) => performUploadPromise(file));
+    while (true) {
+      const completed = yield Promise.all(allUploadPromise);
+      const totallyCompleted = completed.every(async (p) => {
+        try {
+          const done = await p;
+          return done;
+        } catch {
+          return true;
+        }
+      });
```

Comment:
better to not block other uploads, when anyone of them fails... plz use `Promise.allSettled` instead of `Promise.all` :)
```suggestion
      const completed = yield Promise.allSettled(allUploadPromise);
      const totallyCompleted = completed.every((result) => result.status === 'fulfilled');
```